src/ocean/util/container/btree/BTree.d

```diff
@@ -0,0 +1,695 @@
+
+    Params:
+        TreeElementType = type of the element to store.
+        tree_degree = degree of a tree (refer to the documentation)
+        Allocator = allocation technique to use
+
+******************************************************************************/
+
+struct BTree(TreeElementType, int tree_degree, alias Allocator = MallocAllocator,
+        alias TrackingPolicy = TrackAllocationsPolicy)
+{
+    /***********************************************************************
+
+        Recursive inorder iteration. Note that the iteration is undefined
+        in case the tree is changed during the iteration.
+
+        Params:
+            dg = opApply's delegate
+
+        Returns:
+            return value of dg
+
+    ***********************************************************************/
+
+    public int opApply (int delegate (ref ElementType) dg)
+    {
+        return inorder(*this, dg);
+    }
+
+    /**************************************************************************
+
+        Inserts the element in the tree.
+
+        Params:
+            el = element to insert.
+
+        Returns:
+            pointer to a copy of the element in the tree, if the element
+            didn't already exist, null otherwise.
+
+    ***************************************************************************/
+
+    public Const!(ElementType)* insertElement (ElementType el)
+    {
+        if (this.root is null)
+        {
+            this.root = insertNewNode();
+        }
+
+        size_t dummy_index;
+        if (this.impl.search (el, dummy_index))
+        {
+            return null;
+        }
+
+        auto r = this.root;
+        if (this.root.number_of_elements == this.root.elements.length)
+        {
+            auto node = this.impl.insertNewNode();
+            // this is a new root
+            this.root = node;
+            node.is_leaf = false;
+            node.number_of_elements = 0;
+
+            // Old root node is the first child
+            node.child_nodes[0] = r;
+
+            splitChild(node, 0, r);
+            auto ret = this.impl.insertNonFull(node, el);
+            debug (BTreeSanity) check_invariants(*this);
+            if (ret)
+                this.content_version++;
+            return ret;
+        }
+        else
+        {
+            auto ret = this.impl.insertNonFull(this.root, el);
+            debug (BTreeSanity) check_invariants(*this);
+            if (ret)
+                this.content_version++;
+            return ret;
+        }
+    }
+
+    /******************************************************************************
+
+        Deletes the element from the BTree.
+
+        Params:
+            element = element to delete.
+
+        Returns:
+            true if the element was found and deleted, false otherwise.
+
+     ******************************************************************************/
+
+    public bool removeElement (ref ElementType element)
```

Comment:
Yeah, it should be const. What I wanted actually is pass `ref Const!(ElementType)`, but it didn't work in D1 for some reason (I'll reevaluate this), since I don't need a copy of the element, and I don't want to change it. I'll look into normalizing passing convention through the entire module.